### ISSUE TITLE
Parsing .tmx: layer data should be stored as long

### ISSUE THREAD

**Issue by AlmasB** [AlmasB/FXGL#756]
In the format the data is stored as UInt, which java int can't handle and so TMXLevelLoader will fail to parse some flipped tiles

**Comment by adambocco** [comment]
@AlmasB 
Is flipping tiles currently unsupported? Even if a flip is done within the bounds of int, it will still fail here:

https://github.com/AlmasB/FXGL/blob/705e65968326afca51c5be9260b09af97cc388d7/fxgl-entity/src/main/kotlin/com/almasb/fxgl/entity/level/tiled/TilesetLoader.kt#L279-L287

If you haven't started working on this yet, I'll take a look at it and get back to you with a plan.

I read through the reference on flipping tiles [here](https://doc.mapeditor.org/en/stable/reference/tmx-map-format/) and have had success describing the correct orientation with long.

**Comment by AlmasB** [comment]
Flipping *should* be supported, but the parsing is failing due to this issue. You are welcome to have a look and propose a plan, and we can go from there.

**Comment by adambocco** [comment]
My idea is to store Layer.data which is now `List<Int>` to `List<Long>`.

I think something  can be done with base64 encoding with a technique like [this]( http://www.java2s.com/example/java/java.nio/get-unsigned-int-from-bytebuffer.html)

Then the actual flipping can be handled when calling loadView from creatTileLayerEntities like it already is when calling the overloaded loadView createObjectLayerEntities here:

https://github.com/AlmasB/FXGL/blob/b5442a8feff8d9595a5c48b330753bcfc6280e49/fxgl-entity/src/main/kotlin/com/almasb/fxgl/entity/level/tiled/TilesetLoader.kt#L72-L75

But with the addition of a diagonal flip. The 3 most sig. bits can be zeroed by anding with `0x1FFFFFFL` from there to work with `findTileset`.

**Comment by AlmasB** [comment]
> My idea is to store Layer.data which is now List<Int> to List<Long>.

Yeah, I think this is fine.

> I think something can be done with base64 encoding with a technique like this

For now, let's just focus on the raw data (i.e. without base64). So we should probably have a sample to go to fxgl-samples / / sandbox where a tile has 4 states: normal, flip hor, flip ver, flip both. This sample should currently fail to load.

> Then the actual flipping can be handled when calling loadView from creatTileLayerEntities like it already is when calling the overloaded loadView createObjectLayerEntities here:

I assume TilesetLoader won't have to change.

> But with the addition of a diagonal flip. The 3 most sig. bits can be zeroed by anding with 0x1FFFFFFL from there to work with findTileset.

There's some code that deals with object flipping, which seems to be using UInt:

https://github.com/AlmasB/FXGL/blob/b5442a8feff8d9595a5c48b330753bcfc6280e49/fxgl-entity/src/main/kotlin/com/almasb/fxgl/entity/level/tiled/TMXLevelLoader.kt#L395-L404

Would it work in a similar way with tiles? UInt in Kotlin may work for `data` (`List<UInt>`), but I wonder if we should stick to Long to make it easier to interop with Java.

**Comment by adambocco** [comment]
>Would it work in a similar way with tiles? UInt in Kotlin may work for data (List<UInt>), but I wonder if we should stick to Long to make it easier to interop with Java.

Yes it would work either way, I just went with Long because of Java and because the issue called for it.  However, kotlin Uint would save space and if you want I can go with that until I see some conflict with Java.

> For now, let's just focus on the raw data (i.e. without base64). So we should probably have a sample to go to fxgl-samples / / sandbox where a tile has 4 states: normal, flip hor, flip ver, flip both. This sample should currently fail to load.

I can definitely can create a sample, I would create something like this anyway to verify the code. 

>I assume TilesetLoader won't have to change.

For tiles, we are creating a single ImageView by using WritableImage to place the tiles in the correct place in `TilesetLoader.loadView()`.  I was thinking we could do something like this to rotate the image in place:

```           
for (i in 0.. tileWidth ) {
  for (j in 0.. tileHeight) {
    buffer.pixelWriter.setArgb(i, j, sourceImage.pixelReader.getArgb( rotatedPixelX, rotatedPixelY))
  }
}
```

This would require a change to TilesetLoader.

**Comment by AlmasB** [comment]
I think we can move to implementation.

`Long` is fine, let's keep it to avoid potential issues.

I've skimmed through loadView() in tileset loader and I think the flip check needs to be done here:

https://github.com/AlmasB/FXGL/blob/b5442a8feff8d9595a5c48b330753bcfc6280e49/fxgl-entity/src/main/kotlin/com/almasb/fxgl/entity/level/tiled/TilesetLoader.kt#L92

As for the actual flipping, I think it's a useful general method we can have in `Images.kt` (where all image manipulation code lives). `fxgl-entity` reads `fxgl-media` (where Images file lives), so we can use it from there. I'm happy to implement:

```
fun flipHorizontally(Image): Image
fun flipVertically(Image): Image
fun flipDiagonally(Image): Image
```

This will be slower than rotating the image in place (since we effectively iterate twice), but tmx loading doesn't run in critical section of game code, so it's fine.

It would be a good time to start thinking about base64 too. See if you can make it work with Long -- existing tests should be of use if something goes wrong.

**Comment by adambocco** [comment]
@AlmasB  

Hi thanks for the implementation, I have started on this tonight and noticed this:

![image](https://user-images.githubusercontent.com/54448507/110222728-676e0f80-7ea2-11eb-99c8-68a0c77ea420.png)

Diagonal flips need something like this:

```
        val srcX = it.y
        val srcY = it.x
        val color = image.pixelReader.getColor(srcX, srcY)
```

This of course is tricky because of non-square tiles going out of bounds of the image. Tiled has a weird way of handling this, where a diagonally flipped rectangular tile will be present in the neighboring grid-space, although the CSV in the .tmx files still shows 0 in that neighboring tile. 

The new function flipDiagonally() maps pixels the image passed to it and would fail currently on tiles of non-uniform dimensions. To match how Tiled operates, I would need it to return a new image with the X and Y dimensions flipped.
If this change was implemented, I can make a check on every diagonally flipped tile to see if it will go out of bounds and crop the subTexture.

**Edit:**
As I was exploring, I tried creating a map of 16x32 tiles and using 32x16 tiles which failed when placing tiles at the border like this: 
![image](https://user-images.githubusercontent.com/54448507/110225930-0d784480-7eb8-11eb-99a7-c5ed6805452c.png)

with this tile:
![image](https://user-images.githubusercontent.com/54448507/110225934-1ec15100-7eb8-11eb-939d-3c4b3b1b8af8.png)

This map has no flipped tiles but will still fail in TilesetLoader as it tries to write out of bounds. I thought it was worth mentioning because going forward, I'm not sure how you want me to handle cases in which there are non-uniform tile dimensions/map dimensions.

**Comment by AlmasB** [comment]
1. Regarding `flipDiagonally()`, for now let's just read the bits, but not actually do anything with the result:

```
val isFlipDiagonally = ...
if (isFlipDiagonally ) {
   // empty
}
```

We can solve this in the future, as needed.

2. > This map has no flipped tiles but will still fail in TilesetLoader as it tries to write out of bounds. I thought it was worth mentioning because going forward, I'm not sure how you want me to handle cases in which there are non-uniform tile dimensions/map dimensions.

Well spotted, I think we should open a separate issue and not consider out-of-bounds maps that come from Tiled in the context of this one. Please can you open a bug issue.

**Comment by AlmasB** [comment]
This is now closed by #986 and includes horizontal and vertical flipping. Diagonal flipping is not currently supported and an issue will be raised in the future based on demand.
